cpp/src/cylon/mapreduce/mapreduce.hpp

```diff
@@ -0,0 +1,113 @@
+//
+// Created by niranda on 11/22/21.
+//
+
+#ifndef CYLON_CPP_SRC_CYLON_MAPREDUCE_MAPREDUCE_HPP_
+#define CYLON_CPP_SRC_CYLON_MAPREDUCE_MAPREDUCE_HPP_
+
+#include <arrow/api.h>
+#include <utility>
+
+#include "cylon/table.hpp"
+#include "cylon/compute/aggregate_kernels.hpp"
+
+namespace cylon {
+namespace mapred {
+
+struct MapToGroupKernel {
+  virtual ~MapToGroupKernel() = default;
+  virtual Status Map(const arrow::ArrayVector &arrays,
+                     std::shared_ptr<arrow::Array> *local_group_ids,
+                     std::shared_ptr<arrow::Array> *local_group_indices,
+                     int64_t *local_num_groups,
+                     arrow::MemoryPool *pool) const;
```

Comment:
missing  = arrow::default_memory_pool()

Reply:
or should we avoid that in line 30?

Reply:
I think virtual functions can not have default values.